File: scripts/test_pattern/test_composite.py
"""Tests for the Composite pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from composite import File, Directory


class TestComposite(unittest.TestCase):

    # ------------------------------------------------------------------
    # Leaf (File)
    # ------------------------------------------------------------------
    def test_file_returns_size(self):
        self.assertEqual(1024, File("doc.txt", 1024).size())

    def test_file_is_not_directory(self):
        self.assertFalse(File("img.png", 512).is_directory())

    def test_file_name(self):
        self.assertEqual("readme.md", File("readme.md", 256).name)

    # ------------------------------------------------------------------
    # Directory
    # ------------------------------------------------------------------
    def test_empty_directory_size_is_zero(self):
        self.assertEqual(0, Directory("empty").size())

    def test_directory_is_directory(self):
        self.assertTrue(Directory("home").is_directory())

    def test_directory_size_is_child_sum(self):
        d = Directory("root")
        d.add(File("a.txt", 100))
        d.add(File("b.txt", 200))
        self.assertEqual(300, d.size())

    def test_nested_directory_sizes(self):
        inner = Directory("inner")
        inner.add(File("x.cpp", 500))
        inner.add(File("y.cpp", 300))

        outer = Directory("outer")
        outer.add(File("main.cpp", 200))
        outer.add(inner)

        self.assertEqual(1000, outer.size())

    def test_child_count(self):
        d = Directory("src")
        d.add(File("a.cpp", 10))
        d.add(File("b.cpp", 20))
        self.assertEqual(2, len(d.children))

    # ------------------------------------------------------------------
    # Deep nesting
    # ------------------------------------------------------------------
    def test_deep_nested_sizes(self):
        c = Directory("c")
        c.add(File("file.txt", 999))
        b = Directory("b")
        b.add(c)
        a = Directory("a")
        a.add(b)
        root = Directory("root")
        root.add(a)
        self.assertEqual(999, root.size())


if __name__ == "__main__":
    unittest.main()

File: test/pattern/test_composite.cpp
#include "pattern/composite.hpp"
#include <gtest/gtest.h>

using namespace pattern;

// ---------------------------------------------------------------------------
// Leaf (File) tests
// ---------------------------------------------------------------------------
TEST(Composite, FileReturnsItsSize)
{
    File f("doc.txt", 1024);
    EXPECT_EQ(1024LL, f.size());
}

TEST(Composite, FileIsNotDirectory)
{
    File f("img.png", 512);
    EXPECT_FALSE(f.isDirectory());
}

TEST(Composite, FileHasCorrectName)
{
    File f("readme.md", 256);
    EXPECT_EQ("readme.md", f.name());
}

// ---------------------------------------------------------------------------
// Composite (Directory) tests
// ---------------------------------------------------------------------------
TEST(Composite, EmptyDirectorySizeIsZero)
{
    Directory d("empty");
    EXPECT_EQ(0LL, d.size());
}

TEST(Composite, DirectoryIsDirectory)
{
    Directory d("home");
    EXPECT_TRUE(d.isDirectory());
}

TEST(Composite, DirectorySizeIsChildSum)
{
    Directory d("root");
    d.add(std::unique_ptr<FileSystemNode>(new File("a.txt", 100)));
    d.add(std::unique_ptr<FileSystemNode>(new File("b.txt", 200)));
    EXPECT_EQ(300LL, d.size());
}

TEST(Composite, NestedDirectorySizes)
{
    auto inner = std::unique_ptr<Directory>(new Directory("inner"));
    inner->add(std::unique_ptr<FileSystemNode>(new File("x.cpp", 500)));
    inner->add(std::unique_ptr<FileSystemNode>(new File("y.cpp", 300)));

    Directory outer("outer");
    outer.add(std::unique_ptr<FileSystemNode>(new File("main.cpp", 200)));
    outer.add(std::move(inner));

    EXPECT_EQ(1000LL, outer.size());
}

TEST(Composite, ChildCountCorrect)
{
    Directory d("src");
    d.add(std::unique_ptr<FileSystemNode>(new File("a.cpp", 10)));
    d.add(std::unique_ptr<FileSystemNode>(new File("b.cpp", 20)));
    EXPECT_EQ(2u, d.childCount());
}

TEST(Composite, ChildAtReturnsCorrectNode)
{
    Directory d("src");
    d.add(std::unique_ptr<FileSystemNode>(new File("first.cpp", 10)));
    EXPECT_EQ("first.cpp", d.childAt(0)->name());
}

TEST(Composite, ChildAtOutOfRangeThrows)
{
    Directory d("src");
    EXPECT_THROW(d.childAt(0), std::out_of_range);
}

// ---------------------------------------------------------------------------
// Deep nesting
// ---------------------------------------------------------------------------
TEST(Composite, DeepNestedSizes)
{
    // root/a/b/c/file.txt = 999
    auto c = std::unique_ptr<Directory>(new Directory("c"));
    c->add(std::unique_ptr<FileSystemNode>(new File("file.txt", 999)));

    auto b = std::unique_ptr<Directory>(new Directory("b"));
    b->add(std::move(c));

    auto a = std::unique_ptr<Directory>(new Directory("a"));
    a->add(std::move(b));

    Directory root("root");
    root.add(std::move(a));

    EXPECT_EQ(999LL, root.size());
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
